### PR TITLE
feat(core): add option to suppress indentation in templates

### LINKED ISSUE #18694: (cdk cli): Create Smaller CloudFormation Templates
### Description

Allow options when deploying using the cdk cli that would result in physically smaller CloudFormation template.

Currently the template that is generated and deployed is formatted as JSON with whitespace.

From our testing, simply removing the whitespace can result in a template that is roughly 50% of the size. 

Outputting the template as YAML also decreases the template size by roughly the same amount.

### Use Case

We have a CDK project that has several lambda functions behind an API. We just ran into the limit where a CloudFormation template has to be less than 1MB. 

The number of resources we have is relatively low, but due to the auto generated names of resources the template size is quite large. 

If we could minify the template before deploying, or if we could have CDK deploy the template as YAML, we would be able to halve the size of the template and deploy twice as many resources.

### Proposed Solution

This would be a combination of previously proposed solutions.

https://github.com/aws/aws-cdk/issues/8712
The above issue suggests having an option to compact the whitespace. This seems like a reasonable proposal.

https://github.com/aws/aws-cdk/issues/2965
The above issue mentions including the option of writing YAML to `cdk.out`. This appears to have been rejected on the premise that the folder is for machines to read, and that the only benefit of YAML is that it's human readable.

This ignores the fact that the YAML format results in a significantly smaller file size, and that AWS CloudFormation has a hard limit on template file sizes.

### Other information

_No response_

### Acknowledge

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by jesterhazy**
Fixes #18694, #8712 

This change adds an option to suppress indentation in CloudFormation template files. Suppressing indentation will reduce the file size of templates.

Indentation can be set by enabling for specific Stacks using the new `suppressTemplateIndentation` property in `StackProps`, or globally using the new `@aws-cdk/core:suppressTemplateIndentation` context key.

This PR provides additional template size reduction beyond the indentation change in  #19656.

@rix0rrr @mackalex @PatMyron 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by jesterhazy** [comment]
note: integration tests have NOT been run

**Comment by jesterhazy** [comment]
Exemption Request: this change doesn't seem to warrant an integration test.

**Comment by jesterhazy** [comment]
Clarification Request: which readme file would you expect to be update for this?

File: packages/aws-cdk-lib/core/lib/stack.ts [hunk 1]
```diff
@@ -1046,7 +1073,8 @@ export class Stack extends Construct implements ITaggable {
         Annotations.of(this).addInfo(`Number of resources: ${numberOfResources} is approaching allowed maximum of ${this.maxResources}`);
       }
     }
-    fs.writeFileSync(outPath, JSON.stringify(template, undefined, 1));
+    const indent = this._suppressTemplateIndentation ? undefined : 1;
+    fs.writeFileSync(outPath, JSON.stringify(template, undefined, indent));
```
Comment (corymhall) [inline packages/aws-cdk-lib/core/lib/stack.ts:1077]: Can we check the size of the template and if it is too large throw a warning that you should set the `suppressTemplateIndentation` flag or split the resources between stacks.

File: packages/aws-cdk-lib/core/test/stack.test.ts
Comment (corymhall) [inline packages/aws-cdk-lib/core/test/stack.test.ts:1]: Can you add another test that asserts the output is still indented when nothing is set? It doesn't look like an existing test covers that case.

**Review by corymhall — CHANGES_REQUESTED** [review]
@jesterhazy actually after looking into this more I think you are on the right track. It looks like this is the best that we can do (for some reason I was thinking there was still something more we could do). 

I only have 2 minor comments and then I'm good with this.

**Comment by jesterhazy** [comment]
@corymhall what needs to happen to move this forward?

File: packages/aws-cdk-lib/core/lib/stack.ts [hunk 2]
```diff
@@ -1047,7 +1075,20 @@ export class Stack extends Construct implements ITaggable {
         Annotations.of(this).addInfo(`Number of resources: ${numberOfResources} is approaching allowed maximum of ${this.maxResources}`);
       }
     }
-    fs.writeFileSync(outPath, JSON.stringify(template, undefined, 1));
+
+    const indent = this._suppressTemplateIndentation ? undefined : 1;
+    const templateData = JSON.stringify(template, undefined, indent);
+
+    if (templateData.length > TEMPLATE_BODY_MAXIMUM_SIZE) {
+      throw new Error(`Template size for stack '${this.node.path}' exceeds limit: ${templateData.length}/${TEMPLATE_BODY_MAXIMUM_SIZE}`);
```
Comment (corymhall) [inline packages/aws-cdk-lib/core/lib/stack.ts:1083]: Lets change this to a warning too. It will still fail fast on deployment.

**Review by corymhall — DISMISSED** [review]
Can you also add a simple integration test (or update an existing one) that sets the `suppressTemplateIndentations: true`. I just want to have a test that actually deploys a template without indentation.

File: packages/aws-cdk-lib/core/lib/stack.ts (hunk 2 above)
Comment (jesterhazy) [reply packages/aws-cdk-lib/core/lib/stack.ts:1083]: ok

**Review by corymhall — APPROVED** [review]

**Pull request merged**